File: examples/benchmarks/benchmark.cpp
#include <unistd.h>
#include <getopt.h>
#include <sys/time.h>
#include <sys/wait.h>

#include <algorithm>
#include <cstring>
#include <iostream>
#include <cstring>
#include <cassert>
#include <cctype>
#include <cstdlib>
#include <cstdio>
#include <vector>

#include "yampl.h"
#include "yampl/utils/Thread.h"

using namespace yampl;
using namespace std;

static struct timeval start_time, end_time;
static long seconds, useconds;

inline void deallocator(void *, void*){}

void start_clock(){
  gettimeofday(&start_time, NULL);
}

long stop_clock(){
  gettimeofday(&end_time, NULL);
  seconds = end_time.tv_sec - start_time.tv_sec;
  useconds = end_time.tv_usec - start_time.tv_usec;
  return ((seconds) * 1000000 + useconds) + 0.5;
}

Context parseContext(const char *c, string &channelName){
  if(strcasecmp(c, "LOCAL") == 0){
    cout << "context: LOCAL" << endl;
    return LOCAL;
  }else if(strcasecmp(c, "THREAD") == 0){
    cout << "context: THREAD" << endl;
    return THREAD;
  }else if(strcasecmp(c, "DISTRIBUTED") == 0){
    cout << "context: DISTRIBUTED" << endl;
    channelName = "127.0.0.1:3333";
    return DISTRIBUTED;
  }

  exit(-1);
}

Channel parseChannel(string const& channelName, string const& socketImpl, Context ctx)
{
    Channel channel;

    channel.name = channelName;

    string impl = socketImpl;
    transform(begin(impl), end(impl), begin(impl), ::tolower);

    switch (ctx)
    {
        case LOCAL:
            if (impl == "pipe")
                goto LOCAL_PIPE_;
            else if (impl == "shm")
                goto LOCAL_SHM_;
            else
                channel.context = ctx;
            break;
        case LOCAL_SHM:
        LOCAL_SHM_:
            channel.context = LOCAL_SHM;
            break;
        case LOCAL_PIPE:
        LOCAL_PIPE_:
            channel.context = LOCAL_PIPE;
            break;
        case THREAD:
        case DISTRIBUTED:
            channel.context = ctx;
            break;
        default:
            exit(-1);
    }

    return channel;
}

void dumpImpl(const char *impl){
  if(strcasecmp(impl, "pipe") == 0){
    cout << "implementation: PipeSocket" << endl;
  }else if(strcasecmp(impl, "shm") == 0){
    cout << "implementation: SHMSocket" << endl;
  }else{
    cout << "implementation: ZMQSocket" << endl;
  }
}

void dumpSemantics(Semantics semantics){
  if(semantics == MOVE_DATA){
    cout << "semantics: MOVE_DATA" << endl;
  }else{
    cout << "semantics: COPY_DATA" << endl;
  }
}

void client(ISocketFactory *factory, Channel channel, Semantics semantics, void *s_buffer, void *r_buffer, unsigned size, unsigned iterations){
  s_buffer = new char[size];
  ISocket *socket = factory->createClientSocket(channel, semantics, deallocator);

  for(size_t i = 0; i < iterations; i++){
    socket->send(s_buffer, size);
    socket->recv(r_buffer, size);
  }

  delete socket;
}

void server(ISocketFactory *factory, Channel channel, Semantics semantics, void *s_buffer, void *r_buffer, unsigned size, unsigned iterations, unsigned multiplicity){
  s_buffer = new char[size];
  ISocket *socket = factory->createServerSocket(channel, semantics, deallocator);

  start_clock();
  for(size_t i = 0; i < iterations * multiplicity; i++){
    socket->recv(r_buffer, size);
#ifndef NDEBUG
    assert(memcmp(s_buffer, r_buffer, size) == 0);
    memset((void *)r_buffer, 0, size);
#endif
    socket->send(s_buffer, size);
  }

  long t = stop_clock();
  cout << "Latency " << t / (2*iterations*multiplicity) << " microseconds"<< endl;
  cout << "Bandwidth " << (size*iterations*multiplicity*2)/t<< " MB/s" << endl;

  for(unsigned i = 0; i < multiplicity; i++){
    int status;
    wait(&status);
  }

  delete socket;
}

int main(int argc, char *argv[])
{
  int opt;
  const char *impl = "zmq";
  const char *cont = "LOCAL";
  unsigned iterations = 1000;
  unsigned size = 1000000;
  unsigned multiplicity = 1;
  char *s_buffer = 0;
  char *r_buffer = 0;
  string channelName = "service";
  Context context = LOCAL;
  Semantics semantics = MOVE_DATA;

  while((opt = getopt(argc, argv, "i:n:s:yc:m:")) != -1){
    switch(opt){
      case 'i':
	impl = strdup(optarg);
	break;
      case 'n':
	iterations = atoi(optarg);
	break;
      case 's':
	size = atoi(optarg);
	break;
      case 'y':
	semantics = COPY_DATA;
	break;
      case 'c':
	cont = strdup(optarg);
	break;
      case 'm':
	multiplicity = atoi(optarg);
	break;
      default:
	fprintf(stderr, "Usage: %s [-i impl] [-y] [-c context] [-n iterations] [-s size]\n", argv[0]);
	exit(EXIT_FAILURE);
    }
  }

  cout << "iterations: " << iterations << endl;
  cout << "buffer size: " << size << endl;
  cout << "multiplicity: " << multiplicity << endl;

  dumpImpl(impl);
  context = parseContext(cont, channelName);
  dumpSemantics(semantics);
  Channel channel = parseChannel(channelName, string(impl), context);
  s_buffer = new char[size];
  r_buffer = new char[size];

  if(channel.context == LOCAL || channel.context == LOCAL_SHM ||
     channel.context == LOCAL_PIPE || channel.context == DISTRIBUTED)
  {
    for(unsigned i = 0; i < multiplicity; i++){
      if(fork() == 0){
	    ISocketFactory *factory = new SocketFactory();
	    client(factory, channel, semantics, s_buffer, r_buffer, size, iterations);

	    delete factory;
	    return 0;
      } else {
	    continue;
      }
    }

    ISocketFactory *factory = new SocketFactory();
    server(factory, channel, semantics, s_buffer, r_buffer, size, iterations, multiplicity);

    delete factory;
  }
  else if(channel.context == THREAD)
  {
    ISocketFactory *factory = new SocketFactory();
    vector<Thread*> threads;

    Thread s(tr1::bind(server, factory, channel, semantics, s_buffer, r_buffer, size, iterations, multiplicity));

    for(unsigned i = 0; i < multiplicity; i++){
      threads.push_back(new Thread(tr1::bind(client, factory, channel, semantics, s_buffer, r_buffer, size, iterations)));
    }

    for(unsigned i = 0; i < multiplicity; i++){
      threads[i]->join();
    }

    s.join();

    delete factory;
  }

  return 0;
}

File: examples/benchmarks/benchmark.py
#!/usr/bin/env python

import sys
import os
import getopt
import math
import time
import thread

import yampl as ym


def run_async(func):
    """Async decorator (https://code.activestate.com/recipes/576684-simple-threading-decorator/)"""
    from threading import Thread
    from functools import wraps

    @wraps(func)
    def async_func(*args, **kwargs):
        func_hl = Thread(target=func, args=args, kwargs=kwargs)
        func_hl.start()
        return func_hl

    return async_func


def get_tid():
    return int(thread.get_ident()) * 31 % 3**10


def run_fork(func):
    """Fork decorator"""
    import os
    from functools import wraps

    @wraps(func)
    def fork_func(*args, **kwargs):
        pid = os.fork()

        if pid == 0:
            func(*args, **kwargs)
            sys.exit(0)
        else:
            return pid

    return fork_func


class Benchmark:
    """Class containing client/server benchmark asynchronous methods"""
    def __init__(self, channel, size, iterations, timeout=0):
        self.channel = channel
        self.size = int(size)
        self.iterations = int(iterations)
        self.mailbox = {}
        self.watermark = 'yampl'
        self.timeout = timeout

    @staticmethod
    def parse_context(context):
        context = context.lower()

        if context == 'thread':
            ctx = ym.Context.THREAD
        elif context == 'local_shm':
            ctx = ym.Context.LOCAL_SHM
        elif context == 'local_pipe':
            ctx = ym.Context.LOCAL_PIPE
        elif context == 'local':
            ctx = ym.Context.LOCAL
        elif context == 'distributed':
            ctx = ym.Context.DISTRIBUTED
        else:
            raise ValueError('Invalid context')

        return ctx

    @staticmethod
    def parse_context_s(context):
        if context == ym.Context.THREAD:
            ctx = 'thread'
        elif context == ym.Context.LOCAL_SHM:
            ctx = 'local_shm'
        elif context == ym.Context.LOCAL_PIPE:
            ctx = 'local_pipe'
        elif context == ym.Context.LOCAL:
            ctx = 'local'
        elif context == ym.Context.DISTRIBUTED:
            ctx = 'distributed'
        else:
            ctx = ''

        return ctx

    @staticmethod
    def parse_channel(channel):
        ctx = Benchmark.parse_context_s(channel.context)

        if channel.context == ym.Context.DISTRIBUTED:
            name = '127.0.0.1:3333'
        else:
            name = channel.name

        return name, ctx

    @staticmethod
    def make_payload(size, pattern):
        padding = int(size % len(pattern))
        payload = int(math.floor((size / len(pattern)))) * pattern + (padding * '\x00')
        return payload

    @run_fork
    def client_fork(self):
        self.client_thunk()

    @run_async
    def client_async(self):
        print('[+] Client launched on thread %d' % get_tid())
        self.client_thunk()

    def client_thunk(self):
        result = self.client()
        self.mailbox['client'] = result

    def client(self):
        channel_parsed = self.parse_channel(self.channel)
        socket = ym.ClientSocket(channel_parsed[0], channel_parsed[1])
        s_buffer = self.make_payload(self.size, self.watermark)

        start_tm = time.clock()

        for i in range(self.iterations):
            socket.send_raw(s_buffer)

            if self.channel.context == ym.Context.THREAD:
                size = -1

                while size == -1:
                    r_buffer = socket.try_recv_raw(self.timeout)
                    size = r_buffer[0]
            else:
                r_buffer = socket.recv_raw()

        end_tm = time.clock()

        return 1e6 * (end_tm - start_tm)

    @run_fork
    def server_fork(self, multiplicity):
        self.server_thunk(multiplicity)

    @run_async
    def server_async(self, multiplicity):
        print('[+] Server launched on thread %d' % get_tid())
        self.server_thunk(multiplicity)

    def server_thunk(self, multiplicity):
        result = self.server(multiplicity)
        self.mailbox['server'] = result

    def server(self, multiplicity):
        channel_parsed = self.parse_channel(self.channel)
        socket = ym.ServerSocket(channel_parsed[0], channel_parsed[1])
        s_buffer = self.make_payload(self.size, self.watermark)

        start_tm = time.clock()

        for i in range(self.iterations * multiplicity):
            if self.channel.context == ym.Context.THREAD:
                size = -1

                while size == -1:
                    r_buffer = socket.try_recv_raw(self.timeout)
                    size = r_buffer[0]
            else:
                r_buffer = socket.recv_raw()
            socket.send_raw(s_buffer)

        end_tm = time.clock()
        elapsed_tm = 1e6 * (end_tm - start_tm)

        return elapsed_tm


def main():
    sock_ctx = 'local_pipe'
    sock_channel = 'service'

    semantics = ym.Semantics.COPY_DATA  # MOVE_DATA is unsupported
    iterations = 1e3
    payload_size = 1e6
    multiplicity = 1

    opts, args = getopt.getopt(sys.argv[1:], 'n:s:c:m:')

    # Parse options
    for opt in opts:
        if opt[0] == '-n':
            iterations = int(opt[1])
        elif opt[0] == '-s':
            payload_size = int(opt[1])
        elif opt[0] == '-c':
            sock_ctx = opt[1]
        elif opt[0] == '-m':
            multiplicity = int(opt[1])

    print('[+] Running benchmark with: \n - Context: %s\n - Semantics: %s\n'
          ' - Iterations: %d \n - Payload Size: %d\n - Multiplicity: %d'
          % (sock_ctx.upper(), semantics, iterations, payload_size, multiplicity))

    # Run
    sock_channel_ = ym.Channel(sock_channel, Benchmark.parse_context(sock_ctx))
    ctxs = [ym.Context.LOCAL, ym.Context.LOCAL_PIPE, ym.Context.LOCAL_SHM, ym.Context.DISTRIBUTED]

    if sock_channel_.context in ctxs:
        benchmark = Benchmark(sock_channel_, payload_size, iterations)

        pids = []

        for i in range(0, multiplicity):
            pids.append(benchmark.client_fork())

        benchmark.server_thunk(multiplicity)

        for pid in pids:
            os.waitpid(pid, 0)
    elif sock_channel_.context == ym.Context.THREAD:
        """
        The THREAD context doesn't actually yield the expected results
        due to Python's internal multi-threading policy that implicates
        that only one thread can be holding the GIL (Global Interpreter
        Lock) at any one time. The result is highly degraded performance,
        unlock LOCAL, LOCAL_PIPE and DISTRIBUTED which run on their own 
        processes.
        """
        benchmark = Benchmark(sock_channel_, payload_size, iterations)

        benchmark.server_async(multiplicity)

        tasks = []

        for i in range(multiplicity):
            tasks.append(benchmark.client_async())

        for task in tasks:
            task.join()
    else:
        raise ValueError('Unimplemented context')

    server_tm = benchmark.mailbox['server']

    print('[+] Latency: %.0f uS' % (server_tm / (2 * iterations * multiplicity)))
    print('[+] Bandwidth: %.0f MB/s' % (2 * payload_size * iterations * multiplicity / server_tm))

    return


if __name__ == '__main__':
    try:
        main()
    except ValueError:
        print('Usage: python benchmark.py [-c context] [-n iterations] [-s size]')
